osu.Game.Rulesets.Mania/Mods/ManiaModFadeIn.cs

```diff
@@ -19,5 +30,103 @@ public class ManiaModFadeIn : Mod
         public override double ScoreMultiplier => 1;
         public override bool Ranked => true;
         public override Type[] IncompatibleMods => new[] { typeof(ModFlashlight<ManiaHitObject>) };
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<ManiaHitObject> drawableRuleset)
+        {
+            ManiaPlayfield maniaPlayfield = (ManiaPlayfield)drawableRuleset.Playfield;
+
+            foreach (Column column in maniaPlayfield.Stages.SelectMany(stage => stage.Columns))
+            {
+                HitObjectContainer hoc = column.HitObjectArea.HitObjectContainer;
+                Container hocParent = (Container)hoc.Parent;
+
+                hocParent.Remove(hoc);
+                hocParent.Add(new BufferedContainer
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Children = new Drawable[]
+                    {
+                        hoc,
+                        new LaneCover(0.5f, false)
+                        {
+                            RelativeSizeAxes = Axes.Both
+                        }
+                    }
+                });
+            }
+        }
+
+        private class LaneCover : CompositeDrawable
+        {
+            private readonly Box gradient;
+            private readonly Box filled;
+
+            public LaneCover(float initialCoverage, bool reversed)
+            {
+                Blending = new BlendingParameters
+                {
+                    RGBEquation = BlendingEquation.Add,
+                    Source = BlendingType.Zero,
+                    Destination = BlendingType.One,
+                    AlphaEquation = BlendingEquation.Add,
+                    SourceAlpha = BlendingType.Zero,
+                    DestinationAlpha = BlendingType.OneMinusSrcAlpha
+                };
+                InternalChildren = new Drawable[]
+                {
+                    gradient = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both,
+                        RelativePositionAxes = Axes.Both,
+                        Height = 0.25f
+                    },
+                    filled = new Box
+                    {
+                        RelativeSizeAxes = Axes.Both
+                    }
+                };
+                Coverage = initialCoverage;
+                Reversed = reversed;
+            }
+
+            private float coverage;
+
+            public float Coverage
+            {
+                set
+                {
+                    filled.Height = value;
+                    gradient.Y = reversed ? 1 - value - gradient.Height : value;
+                    coverage = value;
+                }
+            }
+
+            private bool reversed;
+
+            public bool Reversed
```

Comment:
If this is going to be controlled by the scrolling direction coming from settings, it shouldn't be public.

It's ambiguous as to what it means (is the default location at the top or at the bottom?). Instead, I think it would be better to have the position be based on the config value and then `ManiaModFadeIn` would apply `Scale = new Vector2(1, -1)` to invert it (you'd need appropriate anchors origins, `Centre` will work).